altapay.php

```diff
@@ -0,0 +1,2353 @@
+            foreach ($results as $result) {
+                $ccToken = $result['ccToken'];
+            }
+        }
+
+        $api = null;
+        try {
+            $api = new AltapayMerchantAPI($cgConf['altapay_url'], $cgConf['user'], $cgConf['password'], null);
+            $response = $api->login();
+
+            if (!$response->wasSuccessful()) {
+                $resErrMsg = $response->getErrorMessage();
+                $resErrCode = $response->getErrorCode();
+                throw new Exception("Could not login to the Merchant API: " . $resErrMsg, $resErrCode);
+            }
+        } catch (Exception $e) {
+            Logger::addLog($e->getMessage(), 3, $e->getCode(), $this->name, $this->id, true);
+            return array(
+                'success' => false,
+                'result' => 'failure',
+                'message' => 'unable to connect to gateway',
+                'additionalInfo' => $e->getMessage(),
+                'payment_form_url' => false,
+            );
+        }
+        try {
+            $response = $api->createPaymentRequest(
+                $cgConf['terminal'],
+                $cgConf['uniqueid'],
+                $amount,
+                $cgConf['currency'],
+                $cgConf['payment_type'],
+                $customer,
+                $cgConf['cookie'],
+                $cgConf['language'],
+                $callback,
+                $transactionInfo,
+                $this->getOrderLines($cart),
+                null,
+                $ccToken,
+                null,
+                null,
+                null,
+                null,
+                null,
+                $customerCreatedDate
+            );
+
+            if (!$response->wasSuccessful()) {
+                $resErrMsg = $response->getErrorMessage();
+                $resErrCode = $response->getErrorCode();
+                throw new Exception("Could not create the payment request: " . $resErrMsg, $resErrCode);
+            }
+
+            return array(
+                'success' => true,
+                'uniqueid' => $cgConf['uniqueid'],
+                'amount' => $amount,
+                'result' => 'Success',
+                'payment_form_url' => $response->getRedirectURL(),
+            );
+        } catch (Exception $e) {
+            Logger::addLog($e->getMessage(), 3, $e->getCode(), $this->name, $this->id, true);
+            return array(
+                'success' => false,
+                'result' => 'failure',
+                'message' => 'unable to obtain payment form url',
+                'additionalInfo' => $e->getMessage(),
+                'payment_form_url' => false,
+            );
+        }
+    }
+
+    /**
+     * Get the remote name of the terminal associated with
+     * this payment method. Will check if currency matches the remote terminal.
+     * @param bool $terminal_id
+     * @param bool $currency
+     * @return bool|Terminal
```

Comment:
Since false is being used as an error indicator this should be written as:
```suggestion
     * @return Terminal|false
```
Much better would be to rewrite it to return `null` on fail